File: ruby/testcode/fileupload_007.rb
require_relative 'shared'

ALLOWED_MIMES = %w[image/jpeg image/png image/gif application/pdf].freeze

# vuln-code-snippet start ruby_fileupload_mime_only
def upload_mime_check(req)
  upload = req.file('file')
  return BenchmarkResponse.bad_request('type not allowed') unless ALLOWED_MIMES.include?(upload[:content_type])
  original_name = upload[:filename]
  File.write("/uploads/#{original_name}", upload[:data]) # vuln-code-snippet vuln-line ruby_fileupload_mime_only
  BenchmarkResponse.ok('uploaded')
end
# vuln-code-snippet end ruby_fileupload_mime_only

File: rust/testcode/fileupload_007.rs
//! CWE-434: Image upload saved without verifying magic bytes match declared type.

// vuln-code-snippet start testcodeFileupload007
pub fn handle(req: &super::shared::BenchmarkRequest) -> super::shared::BenchmarkResponse {
    let filename = req.param("filename");
    let content_type = req.header("content-type");
    let content = req.body_str();
    if content_type.starts_with("image/") {
        let path = format!("uploads/images/{}", filename);
        let _ = std::fs::write(&path, content.as_bytes()); // vuln-code-snippet target-line testcodeFileupload007
        super::shared::BenchmarkResponse::ok(&format!("Image saved: {}", path))
    } else {
        super::shared::BenchmarkResponse::bad_request("Only images allowed")
    }
}
// vuln-code-snippet end testcodeFileupload007
